File: solutions/golang/1971/solution.go
func validPath(n int, edges [][]int, source int, destination int) bool {
    graph := make(map[int][]int)
    seen := make([]bool, n)
    for _, edge := range edges {
        to, from := edge[0], edge[1]
        graph[to] = append(graph[to], from)
        graph[from] = append(graph[from], to)
    }

    seen[source] = true;
    dfs(source, destination, graph, seen);
    return seen[destination]
}

func dfs(vertex int, destination int, graph map[int][]int, seen []bool) {
    for _, neighbor := range graph[vertex] {
        if !seen[neighbor] {
            seen[neighbor] = true
            if neighbor == destination {
                return
            }
            dfs(neighbor, destination, graph, seen)
        }
    }
}

File: solutions/cpp/1971/solution.cpp
class Solution {
    unordered_map<int, vector<int>> graph;
    vector<bool> visited;
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        visited = vector<bool>(n);
        // Do a pre-processing operation. Transform the array of edges into an adjacency list.
        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        visited[source] = true;
        dfs(source, destination);
        return visited[destination];
    }

    void dfs(int v, int destination) {
        for (const auto& neighbor : graph[v]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                if (neighbor == destination) {
                    return;
                }
                dfs(neighbor, destination);
            }
        }
    }
};
